api/client/client_test.go

```diff
@@ -0,0 +1,487 @@
+package client
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/textileio/go-textile-threads/api"
+	es "github.com/textileio/go-textile-threads/eventstore"
+	"github.com/textileio/go-textile-threads/util"
+)
+
+const modelName = "Person"
+
+const schema = `{
+	"$id": "https://example.com/person.schema.json",
+	"$schema": "http://json-schema.org/draft-07/schema#",
+	"title": "` + modelName + `",
+	"type": "object",
+	"required": ["ID"],
+	"properties": {
+		"ID": {
+			"type": "string",
+			"description": "The entity's id."
+		},
+		"firstName": {
+			"type": "string",
+			"description": "The person's first name."
+		},
+		"lastName": {
+			"type": "string",
+			"description": "The person's last name."
+		},
+		"age": {
+			"description": "Age in years which must be equal to or greater than zero.",
+			"type": "integer",
+			"minimum": 0
+		}
+	}
+}`
+
+type Person struct {
+	ID        string `json:"ID"`
+	FirstName string `json:"firstName,omitempty"`
+	LastName  string `json:"lastName,omitempty"`
+	Age       int    `json:"age,omitempty"`
+}
+
+func createPerson() *Person {
+	return &Person{
+		ID:        "",
+		FirstName: "Adam",
+		LastName:  "Doe",
+		Age:       21,
+	}
+}
+
+func TestNewStore(t *testing.T) {
+	_, clean := server(t)
+	defer clean()
+	client := client(t)
+
+	_, err := client.NewStore()
+	if err != nil {
+		t.Fatalf("failed to create new store: %v", err)
+	}
+}
+
+func TestRegisterSchema(t *testing.T) {
+	_, clean := server(t)
+	defer clean()
+	client := client(t)
+
+	storeID, _ := client.NewStore()
+	err := client.RegisterSchema(storeID, modelName, schema)
+	if err != nil {
+		t.Fatalf("failed to register schema: %v", err)
+	}
+}
+
+func TestStart(t *testing.T) {
+	_, clean := server(t)
+	defer clean()
+	client := client(t)
+
+	storeID, _ := client.NewStore()
+	err := client.RegisterSchema(storeID, modelName, schema)
+	checkErr(t, err)
+	err = client.Start(storeID)
+	if err != nil {
+		t.Fatalf("failed to start: %v", err)
+	}
+}
+
+func TestStartFromAddress(t *testing.T) {
+	_, clean := server(t)
+	defer clean()
+	client := client(t)
+
+	storeID, _ := client.NewStore()
+	err := client.RegisterSchema(storeID, modelName, schema)
+	checkErr(t, err)
+
+	// TODO: figure out how to test this
+	// client.StartFromAddress(storeId, <multiaddress>, <read key>, <follow key>)
```

Comment:
Could use some help figuring out how to create the required inputs to test this.

Reply:
This `StarFromAddress` is to join the Thread (Store) created by another server.
So, say you have Server1 with Client1, and Server2 with Client2.

Client1/Server1 will `Start()` as usual. Then Client2 can join the underlying Thread that Client1/Server1 created, so to operate in the same `Store`. In order to do that, after creating Client1/Server` with `Start()`, we should get the multiaddr and keys from Server1, and use that info to `StartFromAddress`. After that, both clients should be operating on the same _shared Store_.

Reply:
nice. I guess the part I need some details on is how to "get the multiaddr and keys from Server1".

Reply:
This is currently a manual process... some methods from the foldersync app:
- create the full multiaddress: https://github.com/textileio/go-foldersync/blob/master/client.go#L167
- parse it on the other end and use with `StartFromAddress`: https://github.com/textileio/go-foldersync/blob/master/client.go#L141

Reply:
Cool, thanks. I'll use some of that.

Reply:
Made an issue for this https://github.com/textileio/go-textile-threads/issues/144